lib/types/jtd-schema.ts

```diff
@@ -28,11 +31,35 @@ type EnumString<T> = [T] extends [never]
     : T
   : null
 
+/** true if type is a union of string literals */
+type IsEnum<T> = null extends EnumString<Exclude<T, null>> ? false : true
+
 /** true only if all types are array types (not tuples) */
 // NOTE relies on the fact that tuples don't have an index at 0.5, but arrays
 // have an index at every number
-type IsElements<T> = [T] extends [readonly unknown[]]
-  ? undefined extends T[0.5]
+type IsElements<T> = false extends IsUnion<T>
+  ? [T] extends [readonly unknown[]]
+    ? undefined extends T[0.5]
+      ? false
+      : true
+    : false
+  : false
+
+/** true if the the type is a values type */
+type IsValues<T> = false extends IsUnion<Exclude<T, null>>
+  ? TypeEquality<keyof Exclude<T, null>, string>
+  : false
+
+/** true if the type is a properties type */
+type IsProperties<T> = false extends IsUnion<Exclude<T, null>>
+  ? null extends EnumString<keyof Exclude<T, null>>
+    ? false
+    : true
+  : false
+
+/** true if type is discriminator type */
+type IsDiscriminator<T> = true extends IsUnion<Exclude<T, null>>
```

Comment:
It is possibly to further refactor into something like

```
IsRecord<T, Union extends boolean> = Union extends IsUnion<Exclude<T, null>>
   ? null extends EnumString<keyof Exclude<T, null>>
     ? false
     : true
   : false
```

(spent some time looking for the difference between the two:)

Reply:
yeah, I like the brevity!